### PR TITLE
[Whisper] Training

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Git clone this fork: https://github.com/patrickvonplaten/whisper and you can train whisper on all ESC datasets

### CODE REVIEW

File: run_speech_recognition_whisper.py [hunk 1]
```diff
@@ -616,29 +732,42 @@ def is_labels_non_zero(transcription):
...
+        if do_lower_case:
+            pred_str = [x.lower() for x in pred_str]
+            wer_lower = metric_wer.compute(predictions=pred_str, references=label_str)
+            cer_lower = metric_cer.compute(predictions=pred_str, references=label_str)
+        else:
+            wer_lower = None
+            cer_lower = None
+
+        # replace all punctuation in the predicted string
+        normalized_pred_str = [re.sub(r'[,.:;@#?!&$\-\"]+\ *', " ", input_str) for input_str in pred_str]
```
Comment (patrickvonplaten) [inline run_speech_recognition_whisper.py]: Hmm are we really catching all punctuation here?

Comment (sanchit-gandhi) [reply run_speech_recognition_whisper.py]: Anything in the `re.sub` will be replaced: https://regex101.com/r/sXlmtD/2

We could use `string.punctuation` and remove all of these terms:
```python
import string
print(string.punctuation)
keep_apostrophe = string.punctuation.replace("'", "")
print(keep_apostrophe)
```

File: run_speech_recognition_whisper.py [hunk 2]
```diff
@@ -0,0 +1,928 @@
...
+    def __call__(self, input_ids, scores):
+        tokens = input_ids
+        logits = scores
+        if tokens.shape[1] == self.sample_begin:
+            logits[:, self.tokenizer.encode(" ") + [self.tokenizer.eot]] = -np.inf
+        return logits
+
+
+class SuppressTokens:
+    def __init__(self, suppress_tokens):
```
Comment (patrickvonplaten) [inline run_speech_recognition_whisper.py:117]: @arthur this will be a logit processor for generate

File: run_speech_recognition_whisper.py [hunk 3]
```diff
@@ -0,0 +1,944 @@
...
+            wandb.log(metrics)
+            write_wandb_pred(pred_str, label_str, prefix="eval")
+
+    if training_args.do_predict:
+        if not training_args.do_train and not training_args.do_eval and report_to_wandb:
+            # manually init wandb
+            wandb.init(project=data_args.wandb_project, name=training_args.run_name)
+        for split in test_split:
+            predict_results = trainer.predict(
+                vectorized_datasets[split], metric_key_prefix=split, logits_processor=logits_processors)
```
Comment (patrickvonplaten) [inline run_speech_recognition_whisper.py:905]: @sanchit-gandhi - note that only for the final predict and eval runs do I add the logits_processors above. If you look into the "eval/wer" chart of the training runs: https://wandb.ai/sanchit-gandhi/whisper?workspace=user-patrickvonplaten it's interesting to see that many WER scores goes up significantly right after training for "eval" => this hints at the logits processors not working very well / being useless. We should evaluate without them just to be sure.

File: run_speech_recognition_whisper.py [hunk 4]
```diff
@@ -0,0 +1,955 @@
...
+        # split inputs and labels since they have to be of different lengths
+        # and need different padding methods
+        input_ids = [feature["input_ids"] for feature in features]
+        labels = [feature["labels"] for feature in features]
+
+        # first, pad the audio inputs to max_len
+        input_ids = torch.concat([to_pad_to_mel(input_val)[None, :] for input_val in input_ids])
+
+        # next, append the eos token to our sequence of labels
+        labels = [[self.time_stamp_token_id] + lab + [self.eos_token_id] for lab in labels]
```
Comment (patrickvonplaten) [inline run_speech_recognition_whisper.py:341]: @sanchit-gandhi here the change

Comment (patrickvonplaten) [reply run_speech_recognition_whisper.py:341]: + removed logits_processors 
+ made sure that BOS is correct

File: run_speech_recognition_whisper.py [hunk 5]
```diff
@@ -863,7 +832,13 @@ def _save(self, output_dir: Optional[str] = None, state_dict=None):
             torch.save(self.args, os.path.join(output_dir, "training_args.bin"))
 
     # Define data collator
-    whisper_data_collator = WhisperDataCollatorWithPadding(eos_token_id=tokenizer.eos_token_id)
+    eos = tokenizer.eos_token_id
+    t_stamp = tokenizer("<|notimestamps|>").input_ids[0]
+    whisper_data_collator = WhisperDataCollatorWithPadding(eos_token_id=eos, time_stamp_token_id=t_stamp)
+
+    # make sure model uses 50257 as BOS
+    bos = tokenizer("<|startoftranscript|>").input_ids[0]
```
Comment (patrickvonplaten) [inline run_speech_recognition_whisper.py:810]: @Arthur see here how the prompt should start like in the beginning

File: run_speech_recognition_whisper.py [hunk 6]
```diff
@@ -431,9 +451,10 @@ def main():
         mlp_layer = model.decoder.blocks[block_idx].mlp
         fc1 = mlp_layer[0]
         act_fn = mlp_layer[1]
-        dropout = nn.Dropout(p=model_args.dropout_rate)
+        dropout_1 = nn.Dropout(p=model_args.dropout_rate)
         fc2 = mlp_layer[2]
-        model.decoder.blocks[block_idx].mlp = nn.Sequential(fc1, act_fn, dropout, fc2, dropout)
+        dropout_2 = nn.Dropout(p=model_args.dropout_rate)
```
Comment (patrickvonplaten) [inline run_speech_recognition_whisper.py:430]: @sanchit-gandhi - actually not sure if we should use the same class for dropout
